File: apps/android/app/src/main/java/com/filo/app/ui/AddArticleScreen.kt
package com.filo.app.ui

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.text.KeyboardActions
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.foundation.verticalScroll
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.setValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.input.ImeAction
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.unit.dp
import com.filo.app.Analytics
import com.filo.app.api.ApiClient
import com.filo.app.api.ErrorMessages
import kotlinx.coroutines.launch

@Composable
fun AddArticleScreen(initialUrl: String, onBack: () -> Unit, onSaved: (() -> Unit)? = null) {
    val scope = rememberCoroutineScope()
    var url by remember(initialUrl) { mutableStateOf(initialUrl) }
    var isSubmitting by remember { mutableStateOf(false) }
    var error by remember { mutableStateOf<AppText?>(null) }

    fun submit() {
        if (isSubmitting || url.isBlank()) return
        scope.launch {
            isSubmitting = true
            error = null
            try {
                val saved = ApiClient.importArticle(url.trim())
                Analytics.track(
                    "add_to_reading_list",
                    mapOf("source" to "manual_url", "created" to saved.created),
                )
                onSaved?.invoke()
            } catch (cause: Exception) {
                error = ErrorMessages.forErrorText(cause)
            } finally {
                isSubmitting = false
            }
        }
    }

    Column(Modifier.fillMaxSize()) {
        FiloHeader(tr("記事を追加"), lead = HeaderLead.Back, onLead = onBack)
        Column(
            modifier = Modifier
                .fillMaxSize()
                .verticalScroll(rememberScrollState())
                .padding(PagePadding),
            verticalArrangement = Arrangement.spacedBy(20.dp),
        ) {
            FiloField(tr("記事URL"), hint = tr("URLをリーディングリストに保存します。")) {
                FiloTextField(
                    url,
                    {
                        url = it
                        error = null
                    },
                    placeholder = "https://example.com/article",
                    keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Uri, imeAction = ImeAction.Done, autoCorrectEnabled = false),
                    keyboardActions = KeyboardActions(onDone = { submit() }),
                    modifier = Modifier.fillMaxWidth(),
                )
            }
            Row {
                FiloButton(
                    tr(if (isSubmitting) "保存中…" else "追加"),
                    ::submit,
                    kind = ButtonKind.Primary,
                    icon = FiloIconName.Plus,
                    enabled = !isSubmitting && url.isNotBlank(),
                )
            }
            error?.let { FiloErrorBox(tr(it)) }
        }
    }
}

File: apps/ios/Filo/Sources/AddArticleScreen.swift
import SwiftUI

struct AddArticleScreen: View {
    let onSaved: () -> Void
    @State private var url: String
    @State private var isSubmitting = false
    @State private var errorMessage: String?
    @Environment(\.filoIsDesktop) private var isDesktop

    init(initialUrl: String, onSaved: @escaping () -> Void) {
        self.onSaved = onSaved
        _url = State(initialValue: initialUrl)
    }

    var body: some View {
        FiloPage("記事を追加", showsBack: true) {
            ScrollView {
                VStack(alignment: .leading, spacing: 24) {
                    VStack(alignment: .leading, spacing: 20) {
                        FiloField(label: "記事URL", hint: "URLをリーディングリストに保存します。") {
                            FiloTextField(placeholder: "https://example.com/article", text: $url, isURL: true) {
                                Task { await submit() }
                            }
                        }
                        FiloButton(isSubmitting ? "保存中…" : "追加", icon: .plus, kind: .primary) {
                            Task { await submit() }
                        }
                        .disabled(isSubmitting || url.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty)
                    }
                    if let errorMessage {
                        FiloErrorBox(message: errorMessage)
                    }
                }
                .frame(maxWidth: FiloMetrics.contentWidth, alignment: .leading)
                .frame(maxWidth: .infinity, alignment: .leading)
                .padding(.horizontal, isDesktop ? FiloMetrics.desktopGutter : FiloMetrics.gutter)
                .padding(.top, 24)
                .padding(.bottom, 64)
            }
        }
    }

    private func submit() async {
        guard !isSubmitting else { return }
        isSubmitting = true
        errorMessage = nil
        do {
            let saved = try await APIClient.shared.importArticle(url: url.trimmingCharacters(in: .whitespacesAndNewlines))
            FiloAnalytics.track("add_to_reading_list", parameters: ["source": "manual_url", "created": saved.created])
            onSaved()
        } catch {
            errorMessage = ErrorMessages.message(for: error)
        }
        isSubmitting = false
    }
}
